### ISSUE TITLE
class-string as array key gives false positive with multiple yields in a generator

### ISSUE THREAD

**Issue by owsygavin** [vimeo/psalm#6294]
# Problem statement

In annotating a Generator function with a class-string, if I have more than one `yields` statement, Psalm fails

# Reproduction

https://psalm.dev/r/7d6cac78be

```php
<?php
interface EntityInterface {}
class JobQueueItem implements EntityInterface {}

/** @return \Generator<int, array{0: array<class-string<EntityInterface>, int>}> */
function Works(): iterable {
    yield [
        [JobQueueItem::class => 2]
    ];
}

/** @return \Generator<int, array{0: array<class-string<EntityInterface>, int>}> */
function Fails(): iterable {
    yield [
        [JobQueueItem::class => 2]
    ];
    
        yield [
        [JobQueueItem::class => 4]
    ];
}


/** @return \Generator<int, array{0: array<class-string<EntityInterface>, int>, 1: string}> */
function Works2(): iterable {
    yield [
        [JobQueueItem::class => 2],
        "some string"
    ];
}

/** @return \Generator<int, array{0: array<class-string<EntityInterface>, int>, 1: string}> */
function Fails2(): iterable {
    yield [
        [JobQueueItem::class => 2],
        "some string"
    ];
    
    yield [
        [JobQueueItem::class => 4],
        "some next string"
    ];
}
```

# Expected output

Psalm passes

# Observed output

```
ERROR: InvalidReturnType - 12:13 - The declared return type 'Generator<int, array{0: array<class-string<EntityInterface>, int>}, mixed, mixed>' for Fails is incorrect, got 'Generator<int, array{array{JobQueueItem: 2|4}}, mixed, void>'

ERROR: InvalidReturnType - 32:13 - The declared return type 'Generator<int, array{0: array<class-string<EntityInterface>, int>, 1: string}, mixed, mixed>' for Fails2 is incorrect, got 'Generator<int, array{array{JobQueueItem: 2|4}, "some next string"|"some string"}, mixed, void>'
```
